### ISSUE TITLE
keyboard is non-responsive on startup because it doesn't detect FocusGained unless I switch away (via alt-tab) first.

### ISSUE THREAD

**Issue by correabuscar** [neovide/neovide#1558]
**EDIT:** ignore everything and just jump [here](https://github.com/neovide/neovide/issues/1558#issuecomment-1260681944) for reproduction steps & workaround.

tl;dr: workaround is [here](https://github.com/neovide/neovide/issues/1558#issuecomment-1258077540)

**Describe the bug**

**To Reproduce**
Steps to reproduce the behavior: **EDIT:** a better/automated way is [here](https://github.com/neovide/neovide/issues/1558#issuecomment-1260387370)

1. Follow the log in another terminal via `$ tail -F neovide_rCURRENT.log` (must be in the same directory as `somefile` from below)
2. Run the following from inside a `qterminal`: `$ neovide --nosrgb --noidle --log somefile`
3. observe that any keys you press have no effect inside the neovide window and the log file doesn't notice them, ie. the log file repeats these lines only:
```
TRACE [neovide::window] Settings geometry PhysicalSize { width: 164, height: 56 }
TRACE [neovide::window] Inner size: PhysicalSize { width: 1476, height: 952 }
```
4. Now alt-tab away from neovide and notice the new log lines:
```
TRACE [neovide::window] Settings geometry PhysicalSize { width: 164, height: 56 }
TRACE [neovide::window] Inner size: PhysicalSize { width: 1476, height: 952 }
TRACE [neovide::window] Settings geometry PhysicalSize { width: 164, height: 56 }
TRACE [neovide::window] Inner size: PhysicalSize { width: 1476, height: 952 }
TRACE [neovide::channel_utils] neovide::bridge::ui_commands::UiCommand Parallel(FocusGained)
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::channel_utils] neovide::bridge::ui_commands::UiCommand Parallel(FocusLost)
TRACE [neovide::window] Settings geometry PhysicalSize { width: 164, height: 56 }
TRACE [neovide::window] Inner size: PhysicalSize { width: 1476, height: 952 }
TRACE [neovide::window] Settings geometry PhysicalSize { width: 164, height: 56 }
TRACE [neovide::window] Inner size: PhysicalSize { width: 1476, height: 952 }
TRACE [neovide::window] Settings geometry PhysicalSize { width: 164, height: 56 }
TRACE [neovide::window] Inner size: PhysicalSize { width: 1476, height: 952 }
```

That's right, it finally sees a `FocusGained` before the `FocusLost`, and I think it's due to the lack of seen `FocusGained` that the keyboard isn't responsive. This kind of behaviour doesn't happen on any other program so far. The window is focused, judging by the title bar being blue.

**Expected behavior**
Keyboard works in neovide without the need to first alt-tab away then alt-tab back.

**Screenshots**
See the window has focus, as far as the window manager is concerned, but neovide doesn't seem to know that, as the keyboard is ignored:
![screen-2022-09-26-08-31-12](https://user-images.githubusercontent.com/106607441/192208379-9521eb21-146a-4be5-967d-d692016d28b9.jpg)

**Desktop (please complete the following information):**
 - OS: Linux Gentoo ~amd64
 - Neovide Version 0.10.1
 - Neovim Version v0.7.2

**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**
Maybe later because (unrelated) I just realized that I cannot copy the text seen inside neovide with the mouse, like I could in the (n)vim that runs in the terminal(when `set mouse=` aka no mouse in vim, but neovide needs `set mouse=a` if I want to use scroll) <s>and thus I think I can't use neovide due to this. </s> Well actually there is an acceptable way: https://github.com/neovide/neovide/issues/1204#issuecomment-1033911753 and since MMB paste is already [working](https://github.com/neovide/neovide/issues/534) it's quite good enough for me, thank you very much :) it's actually even better because now it doesn't also copy the spaces at the end of the lines like it would do in the terminal!

**Additional context**
using LXQT with xfwm4 on X11
```
$ X -version

X.Org X Server 1.21.1.4
X Protocol Version 11, Revision 0
Current Operating System: Linux i87k 5.18.19-gentoo-r1-x86_64 #1 SMP Tue Sep 6 08:59:43 CEST 2022 x86_64
Current version of pixman: 0.40.0

$ xfwm4 -V
        This is xfwm4 version 4.16.1 (revision 5f61a84ad) for Xfce 4.16
        Released under the terms of the GNU General Public License.
        Compiled against GTK+-3.24.32, using GTK+-3.24.34.

        Build configuration and supported features:
        - Startup notification support:                 Yes
        - XSync support:                                Yes
        - Render support:                               Yes
        - Xrandr support:                               Yes
        - Xpresent support:                             Yes
        - X Input 2 support:                            Yes
        - Embedded compositor:                          Yes
        - Epoxy support:                                Yes
        - KDE systray proxy (deprecated):               No



LXQt About Version
1.1.0
LXQt Version
1.1.0
Qt
5.15.5
Build type
Release

```

**Comment by correabuscar** [comment]
It doesn't always happen though, but when it does, I definitely don't see a `FocusGained` in the log! ie. `tail -F neovide_rCURRENT.log|grep -i focus`

**Comment by correabuscar** [comment]
I forgot to mention that the mouse always works even if the keyboard doesn't. ie. clicking around does move the cursor.

**Comment by correabuscar** [comment]
oh it's already been reported: https://github.com/neovide/neovide/issues/960

**Comment by correabuscar** [comment]
Actually I'm reopening it because the OP there says `keyboard input doesn't function until clicking the window with a mouse pointer` which is not the case for me: mouse clicking does work but the keyboard is still unresponsive. It seems to be missing the `FocusGained` even for some reason, and I don't know why, but this might be the issue.

**Comment by correabuscar** [comment]
I've attempted some workaround(s) but they didn't make the issue go away, as it seems to really need that real FocusGained event instead of what I did to pretend neovide got the event:

<details>


```patch
diff --git a/src/window/keyboard_manager.rs b/src/window/keyboard_manager.rs
index 2807dbf..0222198 100644
--- a/src/window/keyboard_manager.rs
+++ b/src/window/keyboard_manager.rs
@@ -20,7 +20,7 @@ pub struct KeyboardManager {
     alt: bool,
     prev_dead_key: Option<char>,
     logo: bool,
-    ignore_input_this_frame: bool,
+    //ignore_input_this_frame: bool,
     queued_input_events: Vec<InputEvent>,
 }
 
@@ -32,7 +32,7 @@ impl KeyboardManager {
             alt: false,
             prev_dead_key: None,
             logo: false,
-            ignore_input_this_frame: false,
+            //ignore_input_this_frame: false,
             queued_input_events: Vec::new(),
         }
     }
@@ -45,7 +45,7 @@ impl KeyboardManager {
             } => {
                 // When window is just focused or lost it's focus, ignore keyboard events
                 // that were submitted this frame
-                self.ignore_input_this_frame = true;
+                //self.ignore_input_this_frame = true;
             }
             Event::WindowEvent {
                 event:
@@ -79,7 +79,7 @@ impl KeyboardManager {
             }
             Event::MainEventsCleared => {
                 // If the window wasn't just focused.
-                if !self.should_ignore_input() {
+                //if !self.should_ignore_input() {
                     // And we have a keyboard event this frame
                     for input_event in self.queued_input_events.iter() {
                         let mut next_dead_key = self.prev_dead_key;
@@ -111,11 +111,11 @@ impl KeyboardManager {
                         }
                         self.prev_dead_key = next_dead_key;
                     }
-                }
+                //}
 
                 // Regardless of whether this was a valid keyboard input or not, rest ignoring and
                 // whatever event was queued.
-                self.ignore_input_this_frame = false;
+                //self.ignore_input_this_frame = false;
                 self.queued_input_events.clear();
             }
             _ => {}
@@ -123,8 +123,9 @@ impl KeyboardManager {
     }
 
     fn should_ignore_input(&self) -> bool {
-        let settings = SETTINGS.get::<KeyboardSettings>();
-        self.ignore_input_this_frame || (self.logo && !settings.use_logo)
+        //let settings = SETTINGS.get::<KeyboardSettings>();
+        //self.ignore_input_this_frame || (self.logo && !settings.use_logo)
+        false
     }
 
     fn maybe_get_keybinding(&self, key_event: &KeyEvent) -> Option<String> {
diff --git a/src/window/mod.rs b/src/window/mod.rs
index 0dccb94..df5c3e9 100644
--- a/src/window/mod.rs
+++ b/src/window/mod.rs
@@ -73,6 +73,7 @@ pub struct GlutinWindowWrapper {
     saved_grid_size: Option<Dimensions>,
     size_at_startup: PhysicalSize<u32>,
     maximized_at_startup: bool,
+    first_time_run:bool,
     window_command_receiver: UnboundedReceiver<WindowCommand>,
 }
 
@@ -179,8 +180,10 @@ impl GlutinWindowWrapper {
                 ..
             } => {
                 if focus {
+                    log::trace!("focused3");
                     self.handle_focus_gained();
                 } else {
+                    log::trace!("unfocused3");
                     self.handle_focus_lost();
                 }
             }
@@ -411,6 +414,7 @@ pub fn create_window() {
         saved_inner_size,
         saved_grid_size: None,
         window_command_receiver,
+        first_time_run: true,
     };
 
     let mut previous_frame_start = Instant::now();
@@ -421,6 +425,11 @@ pub fn create_window() {
         Unfocused,
     }
     let mut focused = FocusedState::Focused;
+    //static mut first_time_run:bool=true;
+    log::trace!("focused0");
+    //no effect:
+    //EVENT_AGGREGATOR.send(UiCommand::Parallel(ParallelCommand::FocusGained));
+    //REDRAW_SCHEDULER.queue_next_frame();
 
     event_loop.run(move |e, _window_target, control_flow| {
         // Window focus changed
@@ -430,8 +439,10 @@ pub fn create_window() {
         } = e
         {
             focused = if focused_event {
+                log::trace!("focused1");
                 FocusedState::Focused
             } else {
+                log::trace!("unfocusednotdrawn1");
                 FocusedState::UnfocusedNotDrawn
             };
         }
@@ -452,6 +463,20 @@ pub fn create_window() {
         window_wrapper.handle_window_commands();
         window_wrapper.synchronize_settings();
         window_wrapper.handle_event(e);
+        if window_wrapper.first_time_run {
+            window_wrapper.first_time_run=false;
+            match focused {
+                FocusedState::Focused => {
+                    log::trace!("focused4");
+                    //yeah this doesn't work either, it gets hit, but keyboard input is still dead
+                    window_wrapper.handle_focus_gained();
+                }
+                _ => {
+                    log::trace!("unfocused4");
+                    window_wrapper.handle_focus_lost();
+                }
+            } // match
+        } // if
 
         let refresh_rate = match focused {
             FocusedState::Focused | FocusedState::UnfocusedNotDrawn => {
@@ -468,6 +493,7 @@ pub fn create_window() {
             let dt = previous_frame_start.elapsed().as_secs_f32();
             window_wrapper.draw_frame(dt);
             if let FocusedState::UnfocusedNotDrawn = focused {
+                log::trace!("unfocused2orwtf");
                 focused = FocusedState::Unfocused;
             }
             previous_frame_start = frame_start;
```

</details>

**Comment by correabuscar** [comment]
Well, I see the problem, sort of:
If the focus event isn't received for whatever reason(and this is the actual problem: why isn't it received? is X not sending it? but then, every other program seems to have no problem) then `self.active_window` [isn't set](https://github.com/neovide/winit/blob/3ad338be3d0bfacd7f7044f5733f4c3baefb702e/src/platform_impl/linux/x11/event_processor.rs#L814-L827) in `winit` and [if that's not set](https://github.com/neovide/winit/blob/3ad338be3d0bfacd7f7044f5733f4c3baefb702e/src/platform_impl/linux/x11/event_processor.rs#L1022-L1023), any keyboard events are ignored and [not sent](https://github.com/neovide/winit/blob/3ad338be3d0bfacd7f7044f5733f4c3baefb702e/src/platform_impl/linux/x11/event_processor.rs#L1048-L1050) as `WindowEvent::KeyboardInput` which means they aren't seen by neovide's [keyboard_manager.rs](https://github.com/neovide/neovide/blob/0b6e20c1994c761814eb6a8ad33dea90ade97573/src/window/keyboard_manager.rs#L50-L61)

The raw keyboard events are still seen (but not handled in keyboard_manager.rs) at the caller [here](https://github.com/neovide/neovide/blob/0b6e20c1994c761814eb6a8ad33dea90ade97573/src/window/mod.rs#L142-L143) but this is really irrelevant because those events are always seen even if the window would be unfocused.

The problem remains: why is neovide sometimes not receiving a [`ffi::XI_FocusIn`](https://github.com/neovide/winit/blob/3ad338be3d0bfacd7f7044f5733f4c3baefb702e/src/platform_impl/linux/x11/event_processor.rs#L814) event (at `winit` level) at startup ?

Alternatively, would it make sense to add another way of [setting the active window](https://github.com/neovide/winit/blob/3ad338be3d0bfacd7f7044f5733f4c3baefb702e/src/platform_impl/linux/x11/event_processor.rs#L826-L827) ? Maybe test if our window is in focused state\* and if it is, and if active_window is None then set it to our window, but where and when...

\* and don't depend on events to find out the focused state.

**Comment by correabuscar** [comment]
Now that I made `cargo` work correctly for my locally modified `winit`, it seems that the following change is a good enough for me workaround even with the lack of received FocusIn/FocusGained:
```patch
diff --git a/src/platform_impl/linux/x11/mod.rs b/src/platform_impl/linux/x11/mod.rs
index da954094..2449fbb1 100644
--- a/src/platform_impl/linux/x11/mod.rs
+++ b/src/platform_impl/linux/x11/mod.rs
@@ -248,7 +248,7 @@ impl<T: 'static> EventLoop<T> {
             device_mod_state: Default::default(),
             num_touch: 0,
             first_touch: None,
-            active_window: None,
+            active_window: Some(root),
         };
 
         // Register for device hotplug events
```

I will have to test it by removing all other changes from neovide and winit and will confirm here: TODO

**Comment by correabuscar** [comment]
I've found a better way of detecting when FocusGained isn't received(but works only on non-debug aka without `--nofork` arg, version of neovide, else it won't exit on error):
```bash
$ export RUST_BACKTRACE=0; true; while test "$?"!=0; do echo boom|neovide - ; done
```
this is based on another neovide bug(I haven't _really_ checked(just superficially) if anyone reported it yet) which makes neovide crash if trying to use stdin as a file (but vim/nvim work fine and I use it in real world like: `diff -up rust-1.62.1.ebuild.orig rust-1.62.1.ebuild|nvim -`)

What happens, in the above one-liner, is that neovide will keep crashing like this
```
ERROR [neovide::error_handling] Could not setup setting notifier for cursor_antialiasing: Error sending request 'nvim_command'
thread '<unnamed>' panicked at 'Could not setup setting notifier for cursor_antialiasing: Error sending request 'nvim_command'', src/error_handling.rs:5:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SendError(Parallel(FocusGained))', src/event_aggregator.rs:55:28
```
until it doesn't receive the FocusGained event, in which case the neovide window remains (and if you press a key or mouse, ie. any event, it crashes afterwards) but otherwise waits for an event in this state:
```
ERROR [neovide::error_handling] Could not setup setting notifier for floating_blur_amount_y: Error sending request 'nvim_command'
thread '<unnamed>' panicked at 'Could not setup setting notifier for floating_blur_amount_y: Error sending request 'nvim_command'', src/error_handling.rs:5:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by pspiagicw** [comment]
Can you provide detailed instructions for getting where you are ? It looks confusing for now.

**Comment by correabuscar** [comment]
Sure thing and sorry for all the noise(I was figuring out things along the way):

As normal user do the following to find out if you have the issue in OP:

1. `cd /tmp`
2. `git clone https://github.com/neovide/neovide`
3. `cd neovide`
4. `cargo build`
5. `$ RUST_BACKTRACE=0; true; while test "$?"!=0; do echo boom|./target/debug/neovide - ; done`
this will start&stop neovide multiple times (the stop happens due to another bug that causes neovide to crash) until after a few times of that it doesn't crash anymore and you see the neovide window on screen. This shows that the FocusGained event wasn't received because it would otherwise crash when receiving any event. And if you click the mouse button inside neovide it will crash. But if you try to type anything like press 'a' key to append, it won't crash because the keyboard inputs are ignored.

**Comment by pspiagicw** [comment]
Thanks! I wanted to fix #1488, by your method. But unfortunately it was not fixed. Glad it worked for your case!

**Comment by correabuscar** [comment]
> Thanks! I wanted to fix #1488, by your method. But unfortunately it was not fixed. Glad it worked for your case! Thanks for the detailed instructions.

There's another more crazy variant which also works for my issue, you might want to try out even though your issue seems somewhat different(needing mouse a click), this is the `winit` patch for it(it doesn't need the other one with `Some(root)` from above):
```patch
diff --git a/src/platform_impl/linux/x11/event_processor.rs b/src/platform_impl/linux/x11/event_processor.rs
index f00182df..1b46eb34 100644
--- a/src/platform_impl/linux/x11/event_processor.rs
+++ b/src/platform_impl/linux/x11/event_processor.rs
@@ -823,7 +823,7 @@ impl<T: 'static> EventProcessor<T> {
 
                         self.device_mod_state.update_state(&modifiers, None);
 
-                        if self.active_window != Some(xev.event) {
+                        //if self.active_window != Some(xev.event) {
                             self.active_window = Some(xev.event);
 
                             let window_id = mkwid(xev.event);
@@ -869,7 +869,7 @@ impl<T: 'static> EventProcessor<T> {
                                 &mut self.device_mod_state,
                                 &mut callback,
                             );
-                        }
+                        //}
                     }
                     ffi::XI_FocusOut => {
                         let xev: &ffi::XIFocusOutEvent = unsafe { &*(xev.data as *const _) };
@@ -881,7 +881,7 @@ impl<T: 'static> EventProcessor<T> {
                             .unfocus(xev.event)
                             .expect("Failed to unfocus input context");
 
-                        if self.active_window.take() == Some(xev.event) {
+                        //if self.active_window.take() == Some(xev.event) {
                             let window_id = mkwid(xev.event);
 
                             // Issue key release events for all pressed keys
@@ -904,7 +904,7 @@ impl<T: 'static> EventProcessor<T> {
                                 window_id,
                                 event: Focused(false),
                             })
-                        }
+                        //}
                     }
 
                     ffi::XI_TouchBegin | ffi::XI_TouchUpdate | ffi::XI_TouchEnd => {
@@ -1020,7 +1020,7 @@ impl<T: 'static> EventProcessor<T> {
                     // The regular KeyPress event has a problem where if you press a dead key, a KeyPress
                     // event won't be emitted. XInput 2 does not have this problem.
                     ffi::XI_KeyPress | ffi::XI_KeyRelease => {
-                        if let Some(active_window) = self.active_window {
+                        //if let Some(active_window) = self.active_window {
                             let state = if xev.evtype == ffi::XI_KeyPress {
                                 Pressed
                             } else {
@@ -1031,6 +1031,8 @@ impl<T: 'static> EventProcessor<T> {
 
                             // We use `self.active_window` here as `xkev.event` has a completely different
                             // value for some reason.
+                            self.active_window = Some(xkev.event);
+                            let active_window=xkev.event;
                             let window_id = mkwid(active_window);
 
                             let device_id = mkdid(xkev.deviceid);
@@ -1064,7 +1066,7 @@ impl<T: 'static> EventProcessor<T> {
                                     is_synthetic: false,
                                 },
                             });
-                        }
+                        //}
                     }
 
                     ffi::XI_RawKeyPress | ffi::XI_RawKeyRelease => {
```

It doesn't sense keys when unfocused, even though I thought it might, so that's good. I don't really understand much of what's going on though, obviously.

**Comment by stefanwatt** [comment]
I tried both patches and neither did anything. 
I'm on arch+i3wm.

**Comment by correabuscar** [comment]
> I tried both patches and neither did anything. I'm on arch+i3wm.

Are you on X11 or Wayland?

Hopefully you didn't try them both at the same time, especially if the first one didn't do anything.

I'll try Artix+i3wm in X11 in virtualbox and edit this with the results.

**Comment by stefanwatt** [comment]
`With i3wm, I can reproduce the issue manually, eg. if I make sure the mouse doesn't hover on neovide when it starts up, sometimes it won't accept keyboard inputs, but if I move mouse to hover on it (it then receives the first FocusGained event and) then it starts working`

Maybe I'm experiencing a different bug then. Key presses will always be recognized after switching focus back and forth.
But never when I first launch neovide through a keybinding. It will work if I launch it via rofi or from terminal, but I'm guessing that's because I switch focus (from rofi/terminal to neovide) in those cases.

**Comment by correabuscar** [comment]
> 



> `With i3wm, I can reproduce the issue manually, eg. if I make sure the mouse doesn't hover on neovide when it starts up, sometimes it won't accept keyboard inputs, but if I move mouse to hover on it (it then receives the first FocusGained event and) then it starts working`
> 
> Maybe I'm experiencing a different bug then. Key presses will always be recognized after switching focus back and forth. But never when I first launch neovide through a keybinding. It will work if I launch it via rofi or from terminal, but I'm guessing that's because I switch focus (from rofi/terminal to neovide) in those cases.

Kinda sounds like the same bug, I'm not sure.
What I tried is assign the key binding ctrl+alt+d to starting neovide and regardless of where the mouse is (especially if not over neovide when it start) neovide is always focused and remains focused when it starts. The only thing I notice is that most of the time the Ctrl key remains stuck in neovide, as if neovide thinkgs I'm still keeping ctrl key pressed when irl I'm not, so I've to unstuck it by pressing it (and releasing it) once, otherwise I can't type ":q". 
I do encounter this issue (but not that often) where it seems focused but not accepting any keyboard input until I hover the mouse over it(which I notice it does a quick focus on my terminal then on neovide right when transitioning between the two windows), so this does seem like either of the two patches should work for you. Or maybe I'm wrong and it's a different issue for you. I'm not sure how you're setting your bindings, maybe for me it's different because I'm using LxQt (and I don't know what you're using as a Desktop Environment and if that matters), and I'm assuming you're on X11 (not Wayland) like me, otherwise, it's for sure a different issue. With the first patch (with Some(root)) I can't hit the issue anymore when using the keybinding and mouse is still hovering over a terminal window (so not on the newly started neovide window).

**Comment by stefanwatt** [comment]
> regardless of where the mouse is (especially if not over neovide when it start) neovide is always focused and remains focused when it starts
same for me 

> The onlynew thing I notice is that most of the time the Ctrl key remains stuck in neovide, as if neovide thinkgs I'm still keeping ctrl key pressed when irl I'm not, so I've to unstuck it by pressing it (and releasing it) once, otherwise I can't type ":q".
Ok. That is interesting. It's kind of the same for me. I use super+c to start it and the super key seems to be "stuck". 

>  I don't know what you're using as a Desktop Environment
Nothing. I don't have a DE. Window manager only (i3).
> I'm assuming you're on X11
Yes, correct.

**Comment by stefanwatt** [comment]
I just tried again with same result. Perhaps I'm misunderstanding the instructions.

If I understand correctly you just need to change `active_window` from `None` to `Some(root)` in the winit directory and then update the Cargo.toml in the neovide directory to use the local version of winit. Afterwards update, build, and use the binary in target directory. The one-liners you provide are just for checking if it worked. 
All of that correct?

**Comment by correabuscar** [comment]
> I just tried again with same result. Perhaps I'm misunderstanding the instructions.
> 
> If I understand correctly you just need to change `active_window` from `None` to `Some(root)` in the winit directory and then update the Cargo.toml in the neovide directory to use the local version of winit. Afterwards update, build, and use the binary in target directory. The one-liners you provide are just for checking if it worked. All of that correct?

Yeah, sounds like you're doing it right.
I sometimes make the mistake of running the release build instead of the debug build, but I doubt it's the case for you.
I'm assuming you changed the shortcut/binding to point to the abs path of ./target/debug/neovide too.
So, I'm out of ideas what could be wrong. Possibly you're hitting a different issue? Maybe look at the log and see if neovide is seeing any keys when the input seems to be ignored ? if yes, then you know you're hitting the same new issue that I did when I tried to use the `while` to reproduce it, which I mention at the bottom of [this comment](https://github.com/neovide/neovide/issues/1558#issuecomment-1267996573)

**Comment by stefanwatt** [comment]
Just out of curiousity I tested what happens when I bind neovide to just the letter "c" (instead of super+c).
As it seems that whichever modifier you use for the keybind gets "stuck" this seemed like an interesting experiment.
Lo and behold, the keypresses get recognized immediately.
Now "c" is not a very convenient binding but luckily my keyboard is QMK configurable. So I added a keymap for an unused F-Key (F17) in my keyboard layout and bound that to opening neovide in i3.

Finally I got at least a workaround.

**Comment by jggc** [comment]
I have to admit I did not read every detail of this thread but I think I have more information to add.

I seem to have the same bug when I start Neovide, I must click the window or alt-tab out/in in order to gain proper focus. If I don't I will somethimes be sending weird keys such as ^H^S^M when I press hjkl to move around other times there is no input at all.

The bit I wanted to add is that I also get a half-focus behavior when I switch virtual desktops. If Neovide is focused and I `CTRL+ALT+RIGHT` and come back with `CTRL+ALT+LEFT` neovide won't be properly focused and I have to alt-tab twice  or click the window. Most of the time I get the weird keystrokes when I do this. Right now I have `hj` moving around properly and `kl` doing nothing at all.

I'm on Arch-Linux Gnome XOrg everything up to date.

**Comment by correabuscar** [comment]
@jggc sounds like the Ctrl key is just "stuck" (as far as neovide sees it), so you might just want to press and then depress Ctrl key(ie. press&release) after switching to neovide; if it works great, another neovide bug. Maybe Alt key is also stuck considering how you switch to neovide. If any of the keys work(ie. neovide senses something), it's not a focus thing, I'd say.

<details>
I've stopped using neovide because I'm too lazy to make it open in the same window, or at least the same location as the original terminal window, rather than either in the same place or where the window manager decides to place it, so I just use neovide because it saves me time. I only wanted the cursor effects anyway.
</details>

**Comment by isovector** [comment]
Any movement on this bug? I'm still experiencing it on Arch w/ xmonad.

**Comment by fredizzimo** [comment]
I think this issue is the same as https://github.com/rust-windowing/winit/issues/2841, so we are still waiting for a winit fix.

**Comment by correabuscar** [comment]
tl;dr: appears fixed.

I've stopped using neovide a long time ago but when trying to repro this issue now, I can't use the same steps as before because there's no automatic crashing happening anymore (which is good, some other bug got fixed) and instead it just starts up and will ignore any keys possibly due to passing it the `-` arg to take `stdin` as file input.
`echo boom|./target/debug/neovide --no-fork --log -`
vs the working version:
`echo boom| nvim -`

So without that automation, I would say it appears to have been fixed from the number of runs I've manually tried. Focus seems to be detected on startup like this:
```
tail: 'neovide_rCURRENT.log' has appeared;  following new file
TRACE [neovide::settings] Initial value load failed for cursor_unfocused_outline_width: Error processing request: 1 - 'Key not found: neovide_cursor_unfocused_outline_width')
TRACE [neovide::channel_utils] neovim_handler OptionSet { gui_option: Unknown("mousefocus", Boolean(false)) }
TRACE [neovide::editor] Option set Unknown("mousefocus", Boolean(false))
TRACE [neovide::channel_utils] UIComand Parallel(FocusLost)
TRACE [neovide::channel_utils] UIComand Parallel(FocusGained)
```
and then when I switch away:
`TRACE [neovide::channel_utils] UIComand Parallel(FocusLost)`
So I guess it works. I'd manually run like like 100 times to make sure focus is always caught, but who has the will for that :) I assume it's fixed. Thanks, keep up the good work!
